rust/composefs-oci/src/repo.rs

```diff
@@ -0,0 +1,439 @@
+                        .ok_or_else(|| anyhow::anyhow!("Missing hardlink target"))?;
+                    rustix::fs::linkat(
+                        ctx.layer_root.as_fd(),
+                        &*path,
+                        ctx.layer_root.as_fd(),
+                        &*target,
+                        AtFlags::empty(),
+                    )
+                    .with_context(|| format!("hardlinking {path:?} to {target:?}"))?;
+                    ctx.stats.meta_count += 1;
+                }
+                tar::EntryType::Symlink => {
+                    let target = entry
+                        .link_name()
+                        .context("linkname")?
+                        .ok_or_else(|| anyhow::anyhow!("Missing hardlink target"))?;
+                    rustix::fs::symlinkat(&*target, ctx.layer_root.as_fd(), &*path)
+                        .with_context(|| format!("symlinking {path:?} to {target:?}"))?;
+                    ctx.stats.meta_count += 1;
+                }
+                tar::EntryType::Char | tar::EntryType::Block => {
+                    todo!()
+                }
+                tar::EntryType::Directory => {
+                    fileutils::ensure_dir(ctx.layer_root.as_fd(), &path)?;
+                }
+                tar::EntryType::Fifo => todo!(),
+                o => anyhow::bail!("Unhandled entry type: {o:?}"),
+            }
+        }
+
+        Ok(res)
+    }
+
+    #[context("Unpacking regfile")]
+    fn unpack_regfile<E: std::io::Read>(
+        &self,
+        global_objects: &Dir,
+        ctx: &mut ImportContext,
+        mut entry: tar::Entry<E>,
+        path: &Path,
+    ) -> Result<()> {
+        use rustix::fs::AtFlags;
+        // First, spool the file content to a temporary file
+        let mut tmpfile = TempFile::new(&ctx.tmp_objects).context("Creating tmpfile")?;
+        let wrote_size = std::io::copy(&mut entry, &mut tmpfile)
+            .with_context(|| format!("Copying tar entry {:?} to tmpfile", path))?;
+        tmpfile.seek(std::io::SeekFrom::Start(0))?;
+
+        // Load metadata
+        let header = entry.header();
+        let size = header.size().context("header size")?;
+        // This should always be true, but just in case
+        anyhow::ensure!(size == wrote_size);
+
+        // Compute its composefs digest.  This can be an expensive operation,
+        // so in the future it'd be nice to do this is a helper thread.  However
+        // doing so would significantly complicate the flow.
+        if self.has_verity() {
+            fileutils::reopen_tmpfile_ro(&mut tmpfile).context("Reopening tmpfile")?;
+            composefs::fsverity::fsverity_enable(tmpfile.as_file().as_fd())
+                .context("Failed to enable fsverity")?;
+        };
+        let mut digest = Digest::new();
+        composefs::fsverity::fsverity_digest_from_fd(tmpfile.as_file().as_fd(), &mut digest)
+            .context("Computing fsverity digest")?;
```

Comment:
In the case where the filesystem doesn't support fs-verity, this will re-read the file to compute the fs-verity. We already have fs-verity digest support in libcomposefs that supports streaming, so we could have computed this digest while writing to the tmpfile above.

I guess that would be unnecessary in the case where the target fs supports fs-verity, as then we compute it twice, but we can avoid that if self.has_verity().

Reply:
It's a good point.  I think though a better optimization will be to move the fsverity computation (whether userspace or kernel) into a worker thread pool, so we leave the main thread's job to basically reading from the network and saving data as quickly as possible.  This way we avoid intermixing network and CPU bound work on the same thread.  (OTOH on modern processors with sha256 acceleration maybe it doesn't matter, I haven't measured)

Reply:
Honestly, I don't know what the cost is of sha256 these days. Sometimes modern CPUs are fast enough that doing computation on memory is basically free compared to the memory bandwidth cost, but I don't know if that is true with sha256. And in this case its even compared to i/O as well. If the file is large, then we may actually have to read it back from disk (rather than cache).

I did some tests with sha256sum on my (pretty fast) machine on a 1GB file. cat to /dev/null takes ~270 msec uncached, and ~65 msec cached (real time).  Running sha256sum instead takes 460 msec uncached (although this time varies more), and 440 msec cached. 

I guess we can say that sha256 computation is the more costly operation compared to disk i/o. Cached read+sha is 7x time slower than a cached read. And total sha256 time is almost the same for uncached and cached reads, meaning that the disk read times were small (or at least pipelined) compared to the computation time.

Given the above we might want something like a threadpool that does fixed block size sha56 computation across n_cpu threads. Then we could just feed it block by block, getting parallelization even across a single large file.